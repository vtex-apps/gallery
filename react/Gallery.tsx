import React from 'react'
import GalleryLayout from './GalleryLayout'
import type { GalleryLayoutProps, Slots } from './GalleryLayout'
//@ts-ignore
import GalleryLegacy, {
  GalleryProps as GalleryLegacyProps,
} from './GalleryLegacy'

/*
 * This type receives Slots directly, instead of using the 'slots' prop to do it.
 * Is the equivalent of adding '[key: string]: ComponentType' at the end of GalleryLayoutProps
 * and removing the 'slots' prop.
 */
type GalleryLayoutPropsWithSlots = Omit<GalleryLayoutProps, 'slots'> & Slots

const Gallery: React.FC<GalleryLegacyProps | GalleryLayoutPropsWithSlots> = (
  props
) => {
  if ('layouts' in props && props.layouts.length > 0) {
    const {
      layouts,
      lazyItemsRemaining,
      products,
      showingFacets,
      summary,
      ...slots
    } = props as GalleryLayoutPropsWithSlots

    return (
      <GalleryLayout
        layouts={layouts}
        lazyItemsRemaining={lazyItemsRemaining}
        products={products}
        showingFacets={showingFacets}
        summary={summary}
        slots={slots}
      />
    )
  }

  return <GalleryLegacy {...(props as GalleryLegacyProps)} />
}

export interface Product {
  /** Product's id. */
  productId: string
  /** Product's cache id. */
  cacheId: string
  /** Product's name. */
  productName: string
  /** Product's description. */
  description: string
  /** Product's categories. */
  categories: unknown[]
  /** Product's link. */
  link?: string
  /** Product's link text. */
  linkText: string
  /** Product's brand. */
  brand?: string
  /** Product's SKU items. */
  items: ProductItem[]
}

interface ProductItemReference {
  Value: string
}

interface ProductItemImage {
  /** Images's imageUrl. */
  imageUrl: string
  /** Images's imageTag. */
  imageTag: string
}

interface ProductItemSeller {
  /** Sellers's commertialOffer. */
  commertialOffer: {
    /** CommertialOffer's price. */
    Price: number
    /** CommertialOffer's list price. */
    ListPrice: number
  }
}

interface ProductItem {
  /** SKU's id. */
  itemId: string
  /** SKU's name. */
  name: string
  /** SKU's referenceId. */
  referenceId?: ProductItemReference[]
  /** SKU's images. */
  images: ProductItemImage[]
  /** SKU's sellers. */
  sellers: ProductItemSeller[]
}

export default Gallery
