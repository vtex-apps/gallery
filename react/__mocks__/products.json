[
  {
    "productId": "2",
    "productName": "Porsche 911",
    "link": "https://storecomponents.vtexcommercestable.com.br/porsche-911/p",
    "brand": "Porsche",
    "items": [
      {
        "name": "Porsche 911",
        "referenceId": [
          {
            "Value": "321",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155395/2017-porsche-911-carrera-s-coupe-angular-front.png?v=636583915191800000",
            "imageTag": "<img src=\"~/arquivos/ids/155395-#width#-#height#/2017-porsche-911-carrera-s-coupe-angular-front.png?v=636583915191800000\" width=\"#width#\" height=\"#height#\" alt=\"Porsche-911\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 220000,
              "ListPrice": 390000,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "1",
    "productName": "Ninja 300 - Special Edition",
    "link": "https://storecomponents.vtexcommercestable.com.br/ninja-300/p",
    "brand": "Kawasaki",
    "items": [
      {
        "name": "Special Edition",
        "referenceId": [
          {
            "Value": "1235",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155394/17ninja300_wsb.jpg?v=636583737653170000",
            "imageTag": "<img src=\"~/arquivos/ids/155394-#width#-#height#/17ninja300_wsb.jpg?v=636583737653170000\" width=\"#width#\" height=\"#height#\" alt=\"17ninja300_wsb\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 24000,
              "ListPrice": 28000,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "13",
    "productName": "Samsung S9 Plus",
    "link": "https://storecomponents.vtexcommercestable.com.br/samsung-s9/p",
    "brand": "Samsung",
    "items": [
      {
        "name": "S9 Plus",
        "referenceId": [
          {
            "Value": "87263191",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155408/samsung_galaxy_s9_plus_phone_template_2048x.jpg?v=636592276295770000",
            "imageTag": "<img src=\"~/arquivos/ids/155408-#width#-#height#/samsung_galaxy_s9_plus_phone_template_2048x.jpg?v=636592276295770000\" width=\"#width#\" height=\"#height#\" alt=\"front\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155409/download--1-.jpg?v=636592276669400000",
            "imageTag": "<img src=\"~/arquivos/ids/155409-#width#-#height#/download--1-.jpg?v=636592276669400000\" width=\"#width#\" height=\"#height#\" alt=\"back\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 3600,
              "ListPrice": 4000,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "12",
    "productName": "Google Pixel - 2 XL",
    "link": "https://storecomponents.vtexcommercestable.com.br/google-pixel2/p",
    "brand": "Google",
    "items": [
      {
        "name": "2 XL",
        "referenceId": [
          {
            "Value": "6666",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155406/download.jpg?v=636592257640700000",
            "imageTag": "<img src=\"~/arquivos/ids/155406-#width#-#height#/download.jpg?v=636592257640700000\" width=\"#width#\" height=\"#height#\" alt=\"pixel-2\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155407/thumb_93290_default_big.jpg?v=636592257862070000",
            "imageTag": "<img src=\"~/arquivos/ids/155407-#width#-#height#/thumb_93290_default_big.jpg?v=636592257862070000\" width=\"#width#\" height=\"#height#\" alt=\"pixel-2xl\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 2200,
              "ListPrice": 2900,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "8",
    "productName": "TV LG - 4K 3D",
    "link": "https://storecomponents.vtexcommercestable.com.br/tv-lg/p",
    "brand": "LG",
    "items": [
      {
        "name": "4K 3D",
        "referenceId": [
          {
            "Value": "6217361",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155401/tv-lg1.jpg?v=636592149449670000",
            "imageTag": "<img src=\"~/arquivos/ids/155401-#width#-#height#/tv-lg1.jpg?v=636592149449670000\" width=\"#width#\" height=\"#height#\" alt=\"lg-tv1\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155402/tv-lg2.jpg?v=636592149454870000",
            "imageTag": "<img src=\"~/arquivos/ids/155402-#width#-#height#/tv-lg2.jpg?v=636592149454870000\" width=\"#width#\" height=\"#height#\" alt=\"lg-tv2\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 2000,
              "ListPrice": 2000,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "11",
    "productName": "Microsystem Sony - 100 RMS",
    "link": "https://storecomponents.vtexcommercestable.com.br/microsystem-sony/p",
    "brand": "Samsung",
    "items": [
      {
        "name": "100 RMS",
        "referenceId": [
          {
            "Value": "1293711",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155405/som-sony1.jpg?v=636592149480230000",
            "imageTag": "<img src=\"~/arquivos/ids/155405-#width#-#height#/som-sony1.jpg?v=636592149480230000\" width=\"#width#\" height=\"#height#\" alt=\"sony-som\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 1760,
              "ListPrice": 2000,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "9",
    "productName": "Som LG - 100 RMS",
    "link": "https://storecomponents.vtexcommercestable.com.br/som-lg/p",
    "brand": "LG",
    "items": [
      {
        "name": "100 RMS",
        "referenceId": [
          {
            "Value": "9812983",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155400/som-lg1.jpg?v=636592149442230000",
            "imageTag": "<img src=\"~/arquivos/ids/155400-#width#-#height#/som-lg1.jpg?v=636592149442230000\" width=\"#width#\" height=\"#height#\" alt=\"som-samsung\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 1600,
              "ListPrice": 2000,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "14",
    "productName": "Motorola Celular - Moto X4",
    "link": "https://storecomponents.vtexcommercestable.com.br/motorola-moto-x4/p",
    "brand": "Motorola",
    "items": [
      {
        "name": "Moto X4",
        "referenceId": [
          {
            "Value": "2863153",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155410/34324-5-smartphone-motorola-moto-x4-xt1900-32gb-tela-de-5-2-dual-chip-12-mp-8-mp-octa-core-e-3gb-ram-preto.jpg?v=636592277942030000",
            "imageTag": "<img src=\"~/arquivos/ids/155410-#width#-#height#/34324-5-smartphone-motorola-moto-x4-xt1900-32gb-tela-de-5-2-dual-chip-12-mp-8-mp-octa-core-e-3gb-ram-preto.jpg?v=636592277942030000\" width=\"#width#\" height=\"#height#\" alt=\"34324-5-smartphone-motorola-moto-x4-xt1900-32gb-tela-de-5-2-dual-chip-12-mp-8-mp-octa-core-e-3gb-ram-preto\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155411/detalhe-motorola-moto-x4-xt1900-azul-topazio.jpg?v=636592278294370000",
            "imageTag": "<img src=\"~/arquivos/ids/155411-#width#-#height#/detalhe-motorola-moto-x4-xt1900-azul-topazio.jpg?v=636592278294370000\" width=\"#width#\" height=\"#height#\" alt=\"detalhe-motorola-moto-x4-xt1900-azul-topazio\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155412/download.jpg?v=636592278711800000",
            "imageTag": "<img src=\"~/arquivos/ids/155412-#width#-#height#/download.jpg?v=636592278711800000\" width=\"#width#\" height=\"#height#\" alt=\"download\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155413/images.jpg?v=636592278890530000",
            "imageTag": "<img src=\"~/arquivos/ids/155413-#width#-#height#/images.jpg?v=636592278890530000\" width=\"#width#\" height=\"#height#\" alt=\"images\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155414/moto-x4-both-colors-2.jpg?v=636592279074070000",
            "imageTag": "<img src=\"~/arquivos/ids/155414-#width#-#height#/moto-x4-both-colors-2.jpg?v=636592279074070000\" width=\"#width#\" height=\"#height#\" alt=\"moto-x4-both-colors-2\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155415/smartphone-motorola-moto-x-x4-xt1900-32gb-12-0-mp-2-chips-android-7-1-nougat-3g-4g-wi-fi-photo196948602-12-20-3d.jpg?v=636592279477730000",
            "imageTag": "<img src=\"~/arquivos/ids/155415-#width#-#height#/smartphone-motorola-moto-x-x4-xt1900-32gb-12-0-mp-2-chips-android-7-1-nougat-3g-4g-wi-fi-photo196948602-12-20-3d.jpg?v=636592279477730000\" width=\"#width#\" height=\"#height#\" alt=\"smartphone-motorola-moto-x-x4-xt1900-32gb-12-0-mp-2-chips-android-7-1-nougat-3g-4g-wi-fi-photo196948602-12-20-3d\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 1540,
              "ListPrice": 1900,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "10",
    "productName": "TV Sony - 4K 3D",
    "link": "https://storecomponents.vtexcommercestable.com.br/tv-sony/p",
    "brand": "Google",
    "items": [
      {
        "name": "4K 3D",
        "referenceId": [
          {
            "Value": "210399183",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155403/tv-sony.jpg?v=636592149473930000",
            "imageTag": "<img src=\"~/arquivos/ids/155403-#width#-#height#/tv-sony.jpg?v=636592149473930000\" width=\"#width#\" height=\"#height#\" alt=\"sony-tv1\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          },
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155404/tv-sony2.jpg?v=636592149477570000",
            "imageTag": "<img src=\"~/arquivos/ids/155404-#width#-#height#/tv-sony2.jpg?v=636592149477570000\" width=\"#width#\" height=\"#height#\" alt=\"sony-tv2\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 1330,
              "ListPrice": 1330,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  },
  {
    "productId": "7",
    "productName": "Som Bluetooth Samsung - 100 RMS",
    "link": "https://storecomponents.vtexcommercestable.com.br/som-samsung/p",
    "brand": "Samsung",
    "items": [
      {
        "name": "100 RMS",
        "referenceId": [
          {
            "Value": "9831831",
            "__typename": "vtex_storegraphql_2_2_1_Reference"
          }
        ],
        "images": [
          {
            "imageUrl": "http://storecomponents.vteximg.com.br/arquivos/ids/155397/som-samsung.jpg?v=636592103297600000",
            "imageTag": "<img src=\"~/arquivos/ids/155397-#width#-#height#/som-samsung.jpg?v=636592103297600000\" width=\"#width#\" height=\"#height#\" alt=\"som-samsung\" id=\"\" />",
            "__typename": "vtex_storegraphql_2_2_1_Image"
          }
        ],
        "sellers": [
          {
            "commertialOffer": {
              "Price": 936,
              "ListPrice": 1200,
              "__typename": "vtex_storegraphql_2_2_1_Offer"
            },
            "__typename": "vtex_storegraphql_2_2_1_Seller"
          }
        ],
        "__typename": "vtex_storegraphql_2_2_1_SKU"
      }
    ],
    "__typename": "vtex_storegraphql_2_2_1_Product"
  }
]