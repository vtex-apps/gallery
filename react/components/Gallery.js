import React, { Component } from 'react'
import { PropTypes } from 'prop-types'

import VTEXClasses from '../constants/CSSClasses'
import { productShape } from '../constants/PropTypes'
import GalleryItem from './GalleryItem'

/**
 * Canonical gallery that displays a list of given products.
 */
export default class Gallery extends Component {
  static propTypes = {
    /** Maximum number of items per line. */
    maxItemsPerLine: PropTypes.number.isRequired,
    /** Maximum number of items per page. */
    maxItemsPerPage: PropTypes.number.isRequired,
    /** Products to be displayed. */
    products: PropTypes.arrayOf(productShape),
    /** ProductSummary props. */
    summary: PropTypes.any,
  }

  static defaultProps = {
    maxItemsPerLine: 5,
    maxItemsPerPage: 10,
    products: [],
  }

  getWidthClass() {
    switch (this.props.maxItemsPerLine) {
      case 5:
        return 'w-100 w-50-m w-20-l'
      case 4:
        return 'w-100 w-50-m w-25-l'
      case 3:
        return 'w-100 w-50-m w-third-l'
    }
  }

  render() {
    return (
      <div className={`${VTEXClasses.GALLERY_CLASS} pa3 w-100 flex flex-wrap justify-start`}>
        {this.props.products.map(item => {
          return (
            <div
              key={item.productId}
              className={`${VTEXClasses.GALLERY_ITEM_CLASS} ${this.getWidthClass()} flex mv2 pa1 fl`}
            >
              <GalleryItem item={item} summary={this.props.summary} />
            </div>
          )
        })}
      </div>
    )
  }
}
