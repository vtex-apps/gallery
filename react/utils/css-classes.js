export default {
  MAIN_CLASS: 'vtex-gallery',
  ITEM_CLASS: 'vtex-gallery_item',
}
