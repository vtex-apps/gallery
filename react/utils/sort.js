function sortByName(a, b, asc) {
  const index = asc ? -1 : 1

  if (a.productName < b.productName) {
    return 1 * index
  }
  if (a.productName > b.productName) {
    return -1 * index
  }

  return 0
}

function sortByPrice(a, b, asc) {
  const index = asc ? -1 : 1
  const aPrice = a.items[0].sellers[0].commertialOffer.Price
  const bPrice = b.items[0].sellers[0].commertialOffer.Price

  if (aPrice < bPrice) {
    return 1 * index
  }
  if (aPrice > bPrice) {
    return -1 * index
  }

  return 0
}

export function sortProducts(products, sortValue) {
  let sortFunction
  let asc

  switch (sortValue) {
    case 'sortBy.lowerPrice':
      sortFunction = sortByPrice
      asc = true
      break
    case 'sortBy.higherPrice':
      sortFunction = sortByPrice
      asc = false
      break
    case 'sortBy.nameAZ':
      sortFunction = sortByName
      asc = true
      break
    case 'sortBy.nameZA':
      sortFunction = sortByName
      asc = false
      break
    default:
      sortFunction = sortByPrice
      asc = false
      break
  }

  return [...products].sort((a, b) => sortFunction(a, b, asc))
}
