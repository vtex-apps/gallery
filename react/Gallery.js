import React, { useMemo, useContext } from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import { pluck, splitEvery } from 'ramda'

import { useDevice } from 'vtex.device-detector'
import { ProductListContext } from 'vtex.product-list-context'
import { useResponsiveValue } from 'vtex.responsive-values'
import { useCssHandles } from 'vtex.css-handles'
import { useRuntime } from 'vtex.render-runtime'

import { LAYOUT_MODE } from './components/LayoutModeSwitcher'
import { productShape } from './constants/propTypes'
import withResizeDetector from './components/withResizeDetector'
import GalleryRow from './components/GalleryRow'
import ProductListEventCaller from './utils/ProductListEventCaller'
import SettingsContext from './components/SettingsContext'

import { Spinner } from 'vtex.styleguide'

/** Layout with one column */
const ONE_COLUMN_LAYOUT = 1

const LAZY_RENDER_THRESHOLD = 2

const CSS_HANDLES = ['gallery']

const { ProductListProvider } = ProductListContext
/**
 * Canonical gallery that displays a list of given products.
 */
const Gallery = ({
  products = [],
  mobileLayoutMode = 'normal',
  maxItemsPerRow = {
    desktop: 5,
    tablet: 3,
    phone: 2,
  },
  minItemWidth = 240,
  width,
  summary,
  showingFacets,
  lazyItemsRemaining,
}) => {
  const { isMobile } = useDevice()
  const { trackingId = 'Search result' } = useContext(SettingsContext) || {}
  const handles = useCssHandles(CSS_HANDLES)
  const { getSettings } = useRuntime()
  const responsiveMaxItemsPerRow = useResponsiveValue(maxItemsPerRow)

  const layoutMode = isMobile ? mobileLayoutMode : 'normal'

  const getItemsPerRow = () => {
    const maxItemsPerMinWidth = Math.floor(width / minItemWidth)

    if (isMobile) {
      if (layoutMode === 'normal') {
        return ONE_COLUMN_LAYOUT
      }

      const maxItemsOnMobile =
        maxItemsPerMinWidth >= 2 ? maxItemsPerMinWidth : 2

      return responsiveMaxItemsPerRow <= maxItemsOnMobile
        ? responsiveMaxItemsPerRow
        : maxItemsOnMobile
    }

    return responsiveMaxItemsPerRow <= maxItemsPerMinWidth
      ? responsiveMaxItemsPerRow
      : maxItemsPerMinWidth
  }

  const itemsPerRow = getItemsPerRow() || responsiveMaxItemsPerRow

  const rows = useMemo(() => splitEvery(itemsPerRow, products), [
    itemsPerRow,
    products,
  ])

  const galleryClasses = classNames(
    handles.gallery,
    'flex flex-row flex-wrap items-stretch bn ph1 na4',
    {
      'justify-center': !showingFacets,
      'pl9-l': showingFacets,
    }
  )

  const isLazyRenderEnabled = getSettings('vtex.store')
    ?.enableSearchRenderingOptimization

  return (
    <ProductListProvider listName={trackingId}>
      <div className={galleryClasses}>
        {rows.map((rowProducts, index) => (
          <GalleryRow
            key={index.toString()}
            widthAvailable={width != null}
            products={rowProducts}
            lazyRender={isLazyRenderEnabled && index >= LAZY_RENDER_THRESHOLD}
            summary={summary}
            displayMode={layoutMode}
            rowIndex={index}
            itemsPerRow={itemsPerRow}
            preloadFirstImage={index === 0}
          />
        ))}
        {typeof lazyItemsRemaining === 'number' && lazyItemsRemaining > 0 && (
          <div
            style={{
              width: '100%',
              // Approximate number, just to add scroll leeway
              height: 300 * Math.ceil(lazyItemsRemaining / itemsPerRow),
            }}
            className="flex justify-center pt10"
          >
            <Spinner />
          </div>
        )}
      </div>
      <ProductListEventCaller />
    </ProductListProvider>
  )
}

Gallery.propTypes = {
  /** Container width */
  width: PropTypes.number,
  /** Products to be displayed. */
  products: PropTypes.arrayOf(productShape),
  /** ProductSummary props. */
  summary: PropTypes.any,
  /** Max Items per Row */
  maxItemsPerRow: PropTypes.number,
  /** Layout mode of the gallery in mobile view */
  mobileLayoutMode: PropTypes.oneOf(pluck('value', LAYOUT_MODE)),
  /** Min Item Width. */
  minItemWidth: PropTypes.number,
  showingFacets: PropTypes.bool,
  trackingId: PropTypes.string,
  lazyItemsRemaining: PropTypes.number,
}

export default withResizeDetector(Gallery)
