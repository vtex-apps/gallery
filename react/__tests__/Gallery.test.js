import React from 'react'
import Gallery from '../Gallery'

import { sortProducts } from '../utils/sort'
import products from '../resources/products.json'
import { mountWithIntl, loadTranslation } from 'enzyme-react-intl'

import { getIntlContextInfo } from '../utils/intlHelpers'

describe('<Gallery /> component', () => {
  function getPrice(item) {
    return item.items[0].sellers[0].commertialOffer.Price
  }

  function renderComponent(props = {}) {
    const { context, childContextTypes, locale } = getIntlContextInfo()

    loadTranslation(`../locales/${locale}.json`)

    const component = mountWithIntl(
      <Gallery products={products} {...props} />,
      {
        context,
        childContextTypes,
      },
    )

    return { component }
  }

  /* eslint-disable */
  it('should render the correct component', () => {
    const { component } = renderComponent()

    expect(component).toBeTruthy()
  })

  it('should order products correctly by name', () => {
    const sortingOptions = ['sortBy.nameAZ', 'sortBy.nameZA']

    let list = [...products]

    expect(list[0].productName).toBe(products[0].productName)

    list = sortProducts(list, sortingOptions[0])

    expect(list[0].productName).toBe('Google Pixel - 2 XL')

    list = sortProducts(list, sortingOptions[1])

    expect(list[0].productName).toBe('TV Sony - 4K 3D')
  })

  it('should order products correctly by price', () => {
    const sortingOptions = ['sortBy.higherPrice', 'sortBy.lowerPrice']

    let list = [...products]

    expect(getPrice(list[0])).toBe(getPrice(products[0]))

    list = sortProducts(list, sortingOptions[0])

    expect(getPrice(list[0])).toBe(220000)

    list = sortProducts(list, sortingOptions[1])

    expect(getPrice(list[0])).toBe(936)
  })
})
