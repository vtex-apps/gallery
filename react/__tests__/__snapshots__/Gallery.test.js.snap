// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderBy /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
  />
  <div
    class="gallery flex flex-row flex-wrap items-stretch bn ph1 pl9-l na4"
  >
    <div
      class="galleryItem pa4"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point: product-summary 
      </div>
    </div>
    <div
      class="galleryItem pa4"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point: product-summary 
      </div>
    </div>
    <div
      class="galleryItem pa4"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point: product-summary 
      </div>
    </div>
    <div
      class="galleryItem pa4"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point: product-summary 
      </div>
    </div>
    <div
      class="galleryItem pa4"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point: product-summary 
      </div>
    </div>
    <div
      class="galleryItem pa4"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point: product-summary 
      </div>
    </div>
    <div
      class="galleryItem pa4"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point: product-summary 
      </div>
    </div>
    <div
      class="galleryItem pa4"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point: product-summary 
      </div>
    </div>
  </div>
</DocumentFragment>
`;
